src/AsmResolver/IO/IBinaryStreamWriter.cs

```diff
@@ -130,6 +133,27 @@ public static void WriteBytes(this IBinaryStreamWriter writer, byte[] buffer)
             writer.WriteBytes(buffer, 0, buffer.Length);
         }
 
+#if NETSTANDARD2_1_OR_GREATER || NETCOREAPP2_1_OR_GREATER
+        /// <summary>
+        /// Writes a buffer of data to the stream.
+        /// </summary>
+        /// <param name="writer">The writer to use.</param>
+        /// <param name="buffer">The data to write.</param>
+        public static void WriteBytes(this IBinaryStreamWriter writer, ReadOnlySpan<byte> buffer)
+        {
+            if (writer is ISpanBinaryStreamWriter spanWriter)
+            {
+                spanWriter.WriteBytes(buffer);
+                return;
+            }
+
+            byte[] array = ArrayPool<byte>.Shared.Rent(buffer.Length);
+            buffer.CopyTo(array);
+            writer.WriteBytes(array, 0, buffer.Length);
+            ArrayPool<byte>.Shared.Return(array);
```

Comment:
We should probably wrap this into a `try-finally` construction to ensure the rented array is returned even in the event `CopyTo` or `WriteBytes` fail.

Reply:
`try`/`finally` is actually discouraged for `ArrayPool` because the program state is potentially undefined in the presence of an exception, so something may still be referencing the rented array as part of its corrupted state, resulting in potentially dangerous behavior later on (there have been some discussions in the C# Discord about making this clearer in the documentation).

Failure to call `ArrayPool<T>.Return` is also considered safe, it just means that the GC will reclaim the array and the operation effectively becomes equivalent to `new T[]` instead of renting.

So a `try`/`finally` here would mostly just serve to slow down the method by adding an exception handler.

Reply:
Interesting. Today I learned.

Reply:
Not sure if I fully agree/understand. 

As far as I know, the pool returns instances of arrays while also holding a reference to it. Thus, a clean up of the array only happens if the pool was also cleaned. Thus, the instance would only be reclaimed if the pool was not shared and collected as well. Not returning the array would mean the array would "live forever", even if all objects still holding a reference to this array (other than the pool itself) would be cleaned up.

Additionally, I understand that returning array instances that are potentially still in use due to the exceptional code path that was taken is dangerous. However, in either case, if we assume that the callee (i.e., WriteBytes) does not pass along the array to anywhere else (which I think is a pretty fair assumption given the semantics of this method), the arrays are not used anywhere else anyways, never reaching such an invalid or corrupted state.

Or am I missing something here?

Reply:
https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Buffers/ArrayPool.cs,73

Reply:
If I'm understanding this correctly, it's also not required that the array being returned was rented. It seems valid for a coder to "donate" an array to the pool, though dubious in usefulness and efficiency.

Reply:
> As far as I know, the pool returns instances of arrays while also holding a reference to it. Thus, a clean up of the array only happens if the pool was also cleaned. Thus, the instance would only be reclaimed if the pool was not shared and collected as well. Not returning the array would mean the array would "live forever", even if all objects still holding a reference to this array (other than the pool itself) would be cleaned up.

[The pool itself doesn't hold a reference to an array after it has been rented](https://github.com/dotnet/runtime/blob/d05415704826f0bd084d2c27f99e4dbe74f62464/src/libraries/System.Private.CoreLib/src/System/Buffers/SharedArrayPool.cs#L66). That is done specifically to allow the GC to reclaim arrays that were not returned to the pool (which is considered a safe thing to do, and the BCL does it in several places to avoid these issues).

> Additionally, I understand that returning array instances that are potentially still in use due to the exceptional code path that was taken is dangerous. However, in either case, if we assume that the callee (i.e., WriteBytes) does not pass along the array to anywhere else (which I think is a pretty fair assumption given the semantics of this method), the arrays are not used anywhere else anyways, never reaching such an invalid or corrupted state.

If you know for certain that the code will not be holding onto the array, then it's fine to have a `try`/`finally` yeah (it just slightly pessimizes the codegen of the method and makes inlining much less likely to occur, if at all). Generally it's only really *desirable* to have the `finally` if that condition is met _and_ you expect there to be a lot of exceptions thrown (in a cancellable async method, for example) so that you don't starve the ArrayPool. Since this is being passed along to user code though, *technically* no guarantee can be made (imagine if that user code innocently passed the array along to a `Stream` that held onto the array in the presence of an exception, for example).

> EDIT: I guess there is an argument that because it is virtual anyone could override the method and make it do something that will keep the reference somewhere... Though I feel bad implementations are always going to be bad, regardless of correct/misuse of the pool.

My comments specifically apply to `ArrayPool<T>.Shared` since that's what's being used here, but if the API took an `ArrayPool<T>` as input then this would be a consideration as well, yeah.

Reply:
For some further context, [see this comment](https://github.com/dotnet/runtime/issues/71249#issuecomment-1165584805) (and the surrounding thread in general).

Reply:
> [The pool itself doesn't hold a reference to an array after it has been rented](https://github.com/dotnet/runtime/blob/d05415704826f0bd084d2c27f99e4dbe74f62464/src/libraries/System.Private.CoreLib/src/System/Buffers/SharedArrayPool.cs#L66). That is done specifically to allow the GC to reclaim arrays that were not returned to the pool (which is considered a safe thing to do, and the BCL does it in several places to avoid these issues).

Interesting, TIL. This is the core of my misunderstanding. If the pool does not keep a reference to the array after it is rented, this seems fine.